### ISSUE TITLE
Received GOAWAY from APNs server with "null" reason

### ISSUE THREAD

**Issue by Photoy** [relayrides/pushy#312]
Hi,

We are using pushy 0.7.2 to send out a large number (about 1.5 million tokens) of notifications. When first about 5000 tokens being sent (~3000 succeeded, ~2000 rejected), we get repeated Http2GoAwayException:

```
io.netty.handler.codec.http2.StreamBufferingEncoder$Http2GoAwayException: null
        at io.netty.handler.codec.http2.StreamBufferingEncoder.cancelGoAwayStreams(StreamBufferingEncoder.java:245) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.StreamBufferingEncoder.access$000(StreamBufferingEncoder.java:55) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.StreamBufferingEncoder$1.onGoAwayReceived(StreamBufferingEncoder.java:117) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.DefaultHttp2Connection.goAwayReceived(DefaultHttp2Connection.java:228) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.onGoAwayRead0(DefaultHttp2ConnectionDecoder.java:186) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onGoAwayRead(DefaultHttp2ConnectionDecoder.java:523) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readGoAwayFrame(DefaultHttp2FrameReader.java:564) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:265) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:155) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:113) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:333) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:393) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:406) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:243) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:335) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:327) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:266) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:335) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:327) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1070) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:904) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:406) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:243) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:335) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:327) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1336) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:335) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:911) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:544) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:485) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:399) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:371) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742) [netty-4.1.0.Final.jar:4.1.0.Final]
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145) [netty-4.1.0.Final.jar:4.1.0.Final]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_05]
[18:19:34 WARN ] Sending
java.util.concurrent.ExecutionException: io.netty.handler.codec.http2.StreamBufferingEncoder$Http2GoAwayException
        at io.netty.util.concurrent.AbstractFuture.get(AbstractFuture.java:41) ~[netty-4.1.0.Final.jar:4.1.0.Final]
        at com.idreamsky.push.apns.APNSAgent.operationComplete(APNSAgent.java:101) ~[APNSAgent.jar:na]
```

The code snap is as below:

```java
@Override
public void operationComplete(Future<PushNotificationResponse<SimpleApnsPushNotification>> f) {
	try {
		try {
			PushNotificationResponse<SimpleApnsPushNotification> resp = f.get();
			SimpleApnsPushNotification notification = resp.getPushNotification();
			if (resp.isAccepted()) {
				if (listener != null) {
					listener.onAccepted(appId, notification.getToken());
				}
				LOG.debug("[Sent] {}", notification.getToken());
			} else {
				if (listener != null) {
					listener.onRejected(appId, notification.getToken(), resp.getRejectionReason());
				} else {
					tokenDao.removeToken(appId, notification.getToken());
				}
				LOG.debug("[Rejected] {}: {}", notification.getToken(), resp.getRejectionReason());
			}
		} catch (ExecutionException e) {
			if (listener != null) {
				listener.onError(appId, e.getCause());
			}
			LOG.warn("Sending", e);
			if (e.getCause() instanceof ClientNotConnectedException) {
				apns.getReconnectionFuture().await(1000 * 60 * 30);//30min timeout
			}
		}
	} catch (InterruptedException e) {
		LOG.error(e.getMessage());
	}
}
```

**Comment by jchambers** [comment]
My guess is that there are two things going on here:

1. You're connecting to a development environment with a production certificate or vice versa.
2. The [APNs server is sending two `GOAWAY` frames](http://openradar.appspot.com/radar?id=5059955027607552) in that case. I reported the issue to Apple a while ago, and have reason to believe it will be fixed soon.

Can you confirm that is (or isn't) the case?

**Comment by Photoy** [comment]
Hi jchambers, 

First, thank you for the replay and the best java APNs client API I can get. 

In our case, we do use a production certificate and connect to the production environment. And it will always be OK if we try send to one or a few tokens. Also, in our case, the first 5000 notifications were sent out successfully - either succeeded or being rejected.

And,  do you consider the Http2GoAwayException as a case of  ClientNotConnectedException or our client application should handle this exception in the program?

Regards!

**Comment by jchambers** [comment]
Strange. I'm not sure why the server would be closing your connections after several thousand notifications. Is any reason given when you receive a `GOAWAY` frame? If we receive a `GOAWAY` frame, we should log it at the `INFO` level.

> And, do you consider the Http2GoAwayException as a case of ClientNotConnectedException or our client application should handle this exception in the program?

In short, no. The client is connected (although it will soon disconnect), but streams are being (correctly) terminated because the server sent a `GOAWAY` frame.

**Comment by Photoy** [comment]
> Is any reason given when you receive a GOAWAY frame?

No, the reason part is "null" as you can see on the 1st line of the logging:
_io.netty.handler.codec.http2.StreamBufferingEncoder$Http2GoAwayException: null_

**Comment by jchambers** [comment]
My mistake; it wasn't clear until I read the [source for `Http2GoAwayException`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java#L245) that the `null` was referring to the debug data. Still, I'm suspicious that you might be receiving multiple `GOAWAY` frames. Could you please humor me and check to see if there's any possibility that's the case?

**Comment by maor-teradata** [comment]
Adding my 2 cents:
We did a quick test to try sending 20K notifications, all are invalid tokens, as a test case before using real tokens. We want to run an "invisible" test on a large group of tokens we have.
We are running on production environment with production certificate.
Sending a small (~20) tokens, everything works fine.
When we boost it up to 20K, after approx. 5K we also get exceptions.

When I send a single notification at a time, waiting for it to return before sending the next one, i get multiple exceptions of type:
java.util.concurrent.ExecutionException: io.netty.handler.codec.http2.StreamBufferingEncoder$Http2GoAwayException
                at io.netty.util.concurrent.AbstractFuture.get(AbstractFuture.java:41)
.
.
.
Caused by: io.netty.handler.codec.http2.StreamBufferingEncoder$Http2GoAwayException
                at io.netty.handler.codec.http2.StreamBufferingEncoder.cancelGoAwayStreams(StreamBufferingEncoder.java:243)
                ... 39 more

when I send all the notifications and add a listener to the futures, i get multiple exceptions of type:
java.util.concurrent.ExecutionException: com.relayrides.pushy.apns.ClientNotConnectedException: Client disconnected unexpectedly.
                at io.netty.util.concurrent.AbstractFuture.get(AbstractFuture.java:41)
.
.
.
Caused by: com.relayrides.pushy.apns.ClientNotConnectedException: Client disconnected unexpectedly.
                ... 8 more

**Comment by jchambers** [comment]
> When we boost it up to 20K, after approx. 5K we also get exceptions.

Interesting. To be clear, this is when you send 20k notifications with *known-bad* tokens? I wonder if this is some kind of undocumented spam defense type thing on the server side.

> When I send a single notification at a time, waiting for it to return before sending the next one

I'm not totally sure I understand this scenario. Is this in the "sending 20k notifications" case, or the "fewer than a hundred" case?

**Comment by maor-teradata** [comment]
Yes, all tokens are invalid tokens. We wanted to generate a test and see it work flawlessly before starting to use real data.
I reckon it could be some sort of DoS prevention mechanism, but I'm not sure this would be practical due to situations like the one described above, of 3K valid tokens and 2K invalids.

When I sent 20K messages one at a time, waiting with future.get(), I could see that the exception occurred somewhere above the 5K mark.
When I sent the whole 20K, adding listeners on the futures, I could see from the sysout messages that all 20K were sent, and then I started getting a whole lot of exceptions. I could possibly initially be getting the GoAwayException in this case too, but what I manage to see is the NotConnectedException.

And I guess this could possibly help to clarify this issue: is there a way to send a notification and mark it as a test notification for Apple to not really try and send?
If such an option exists, I can try sending my 5K invalid tokens and see if Apple still cuts the connection

**Comment by maor-teradata** [comment]
Another exception popped up:
WARNING: Failed to mark a promise as failure because it's done already: DefaultChannelPromise@7922cfc0(failure: java.nio.channels.ClosedChannelException)
java.nio.channels.ClosedChannelException

Note: this is when I tried sending 1000 real tokens, of which I do not know how many are good and how many are bad.

**Comment by jchambers** [comment]
> Another exception popped up

We'll need more context to understand what's going on there, but I don't believe it's a problem. In any case, it sounds like a separate issue; if you'd like to pursue it, could you please open a new issue and share additional details there?

**Comment by jchambers** [comment]
> s there a way to send a notification and mark it as a test notification for Apple to not really try and send?

No. You could just send silent push notifications with alert-less payloads, though. They'll be delivered, but won't necessarily trigger an action when they arrive.

**Comment by Photoy** [comment]
> Still, I'm suspicious that you might be receiving multiple GOAWAY frames. Could you please humor me and check to see if there's any possibility that's the case?

Sorry for the late response (just come back from holiday).
Of course, I'd like to investigate this possibility, but since we are using it in a production environment, I need to figure out how to run this kind of test. Or do you have any advice about the test?

**Comment by jchambers** [comment]
> Or do you have any advice about the test?

You should just need to set your log level to `INFO` (or finer) and look for this line: https://github.com/relayrides/pushy/blob/master/pushy/src/main/java/com/relayrides/pushy/apns/ApnsClientHandler.java#L223

**Comment by Photoy** [comment]
> You should just need to set your log level to INFO (or finer) and look for this line: https://github.com/relayrides/pushy/blob/master/pushy/src/main/java/com/relayrides/pushy/apns/ApnsClientHandler.java#L223

Got it! Will check the log later...

More findings:
1) if sending a large number (for example 10k) of notifications continually in an asynchronized way (if I'm right): 

> apnsClient.sendNotification(notification). addListener()

will get the exception mentioned above! Small number (like 1k) is still OK.

2) if in a blocking way:
> apnsClient.sendNotification(notification).get()

will be successfully sent out (arrived or rejected). Even when I try to send 1 million notifications through 128 threads, it is OK - without any error!

Issue of APNs gateway capacity? networking?

**Comment by maor-teradata** [comment]
Updating that we managed to solve our issue, posting it in case it helps Photoy or future readers: 

Since it was a quick P.O.C. test, it turns out that the apnsClient was disconnecting before all the responses were received. When we removed the disconnect(), it received all the responses. (and had to be shut down manually, naturally).

**Comment by jchambers** [comment]
@maor-teradata @Photoy I assumed you two were colleagues; is that not the case?

**Comment by jchambers** [comment]
(or, put differently, should we consider your two cases independently?)

**Comment by maor-teradata** [comment]
Not colleagues, two independent cases.
I saw this issue that was similar to mine, and figured that it would probably be better to group them together rather than open a separate issue over the same symptoms.

**Comment by jchambers** [comment]
Understood; thanks for the clarification.

@Photoy, please do let us know if @maor-teradata's solution works for you!

**Comment by Photoy** [comment]
@jchambers, Yes, it is the case! Verified! @maor-teradata, Thank you.
I assumed that the disconnect() method would handle the "naturally shutdown".

**Comment by jchambers** [comment]
Cool! Glad it's resolved. It sounds like we can do more to clarify the situation in the README, docs, and FAQ. Will close this issue when those are all updated.

**Comment by jchambers** [comment]
Updated the docs/README in 0facb6b and added an entry to the [FAQ](https://github.com/relayrides/pushy/wiki/Frequently-asked-questions). Calling this "resolved" for now; @Photoy, @maor-teradata, please leave a note here if you think there's more we should do and we'll revisit the issue.
